### ISSUE TITLE
Alt + Lab queue behaviour bug?

### ISSUE THREAD

**Issue by Jazcash** [beyond-all-reason/Beyond-All-Reason#1522]
Sometimes it cancels the currently building unit to insert the new unit, and other times it inserts it _after_ the currently building unit has finished. I don't know if this is intentional, and if so, what's the criteria for each behaviour?

Personally, I'm not even sure which I prefer, as it differs based on circumstance. I think it should probably be consistent though, or at least a user setting.

**Comment by Jazcash** [comment]
> It depends if the factory is on repeat or not
> If it's on repeat then it's gonna finish what it's currently making. If it's not on repeat it's gonna cancel
> - Damgam

**Comment by Jazcash** [comment]
I think choosing one behaviour or the other based on the factory's repeat mode is a bit arbitrary

**Comment by WatchTheFort** [comment]
I concur. Which mode is more useful? After the current unit, because you can cancel a unit in progress, but you can't uncancel a unit that's been cancelled?

**Comment by Jazcash** [comment]
I think ideally we should be able to do both. Two common scenarios that happen almost every game:

Ally asks to buy T2 con, you queue it up with the insert command. In this scenario you don't want to cancel the currently building unit. I've done this before and it's cancelled expensive units that are 90% built and it's very frustrating.

Second scenario: Enemy sends bombers and you insert lab queue to build AA units. Here you absolutely want to cancel whatever's building immediately, and having it queue up the AA units _after_ some other expensive unit thats currently only 10% built is very frustrating.

Ideally, for me, Alt + queue would insert the order without cancelling the current unit, and a free key combo such as Alt + Space would insert the order and cancel the current unit.

**Comment by sprunk** [comment]
You know what to do https://github.com/ZeroK-RTS/Zero-K/blob/master/LuaUI/Widgets/gui_chili_integral_menu.lua#L1039

ZK does not implement a "insert but cancel" because you can just cancel the thing manually with a single click, but it's trivial to add that as well if you want to.

**Comment by badosu** [comment]
That should be feasible to change in build/gridmenu without much effort. Question is what is the agreed "improved" behavior.

Personally I don't have a problem with current behavior, but I agree standardizing for appending to queue after current regardless of repeat makes sense, with an optionally bound action for cancel current and start immediately.

**Comment by Jazcash** [comment]
Yeah, if we're choosing one behaviour, then not cancelling the current unit is preferable

**Comment by Ezahiel** [comment]
> Yeah, if we're choosing one behaviour, then not cancelling the current unit is preferable

So you wanna current unit to finish before added one with Alt begin right ? Similar to the one with Factory On Repeat right ?

**Comment by PtaQQ** [comment]
Item approved by the Game Design Team.

**Comment by sprEEEzy** [comment]
If the solution is 1 mode only, then it has to be the non-cancelling one (how it is with Repeat: On currently)

**Comment by badosu** [comment]
### Proposed solution:

- Store buildqueue commands after current item on front of queue
- Add desired unit to buildqueue (after current), respecting cmdopts (shift, ctrl, etc), notice if alt is in cmdopts this widget should do nothing (its removal of unit)
- Reissue previous stored buildqueue commands
- See part of the code for storing the commands and reissuing them for reference [here](https://github.com/beyond-all-reason/Beyond-All-Reason/pull/2194/files#diff-2336abed613b82ef97d3e007ceacc59ee89a996608506825e2e1aca6828f8010R77)

### Challenge with proposed solution:

- Performance decreases with buildqueue size (it does 1-by-1), for repeated buildqueue args one can issue a combination of ctrl and shift cmdopts, see [this](https://gist.github.com/badosu/c656fa654aadc61b99df8db5e6b3c15f#file-factory_ui_keys-lua-L228) for example. While the issue persists, it is mitigated a fair amount

**Comment by sprunk** [comment]
Why store and reissue? Just give an insert command if it's a fac not already on repeat https://github.com/ZeroK-RTS/Zero-K/blob/master/LuaUI/Widgets/gui_chili_integral_menu.lua#L1041-L1045

**Comment by badosu** [comment]
Thank you, updated!

**Comment by sprunk** [comment]
The code i linked above has a loop for ctrl/shift. See https://github.com/beyond-all-reason/spring/pull/1253

**Comment by RogShotz** [comment]
Please implement this 🙏

Id take a crack but only have chobby experience. I might look at this soon though... surely it can't be that hard...

To clarify since I don't think it was said, if someone REALLY wants to cancel something its easy enough for them to right cancel the one boat and requeue it rather then have to go through a toggle rigamaroll.

**Comment by hihoman23** [comment]
I made a PR, which was waiting on an extremely obscure engine bug being fixed. It might have been fixed by now (most likely not the case) but another approach might also work. If you do end up doing it, test it well with the wait command being issued (which is where I experienced the bug).

**Comment by RogShotz** [comment]
> I made a PR, which was waiting on an extremely obscure engine bug being fixed. It might have been fixed by now (most likely not the case) but another approach might also work. If you do end up doing it, test it well with the wait command being issued (which is where I experienced the bug).

I actually tend on fixing waits as well in a separate issue. Unsure how they will work with each other. As it stands waiting a lab does not halt all M and E being fed to it. It only fully waits once the current unit is complete.

**Comment by sparr** [comment]
Idea: While holding alt in non-repeat mode, render an X over the currently building unit, to tell the player "if you click or press a key to build a unit right now, it will cancel the build in progress"

**Comment by sparr** [comment]
#8889 is my attempt to implement my previous comment, with a red highlight instead of an X.
